json.go

```diff
@@ -65,28 +81,63 @@ func (j jsonIndexPart) Keys(document Document) ([]Key, error) {
 	return keys, nil
 }
 
-func (j jsonIndexPart) matchRecursive(parts []string, val interface{}) []interface{} {
+func (j jsonIndexPart) Tokenize(value interface{}) []interface{} {
+	if j.tokenizer == nil {
+		return []interface{}{value}
+	}
+
+	if s, ok := value.(string); ok {
+		tokens := j.tokenizer(s)
+		result := make([]interface{}, len(tokens))
+		for i, t := range tokens {
+			result[i] = t
+		}
+		return result
+	}
+	return []interface{}{value}
+}
+
+func (j jsonIndexPart) Transform(value interface{}) interface{} {
+	if j.transformer == nil {
+		return value
+	}
+	if s, ok := value.(string); ok {
+		return j.transformer(s)
+	}
+	return value
+}
+
+func (j jsonIndexPart) matchRecursive(parts []string, val interface{}) ([]interface{}, error) {
 
 	if a, ok := val.([]interface{}); ok {
 		var ra []interface{}
 		for _, ai := range a {
-			interm := j.matchRecursive(parts, ai)
+			interm, err := j.matchRecursive(parts, ai)
```

Comment:
all these 1 or 2 character variables makes it hard to understand the code (because I don't work on it), since I've got to really analyze it to guess/derive what is actually means

Reply:
part of the refactor.